feat: add option for bold/italic formatting

rathharper33 opened the pull request:
I added the option to format line numbers with bold or italics. I changed the name of your function `M.set_highlight` to `M.set_highlight_and_format` with a new parameter `format`.

Note: I haven't seen the syntax `---@param color string` before, so you might have to update that for the new `format` parameter.

rathharper33 committed:
feat: add option for bold/italic formatting

rathharper33 committed:
feat: add option in README.md

mawkler commented:
@itzaeon Thank you very much for your PR! I have just a couple of minor things to comment on.

I think I might prefer if the configuration API looked like this:

```lua
{
  highlights = {
    modes = {
      ['i']  = {
        color = colors.insert,
        bold = false,
        italic = false,
      },
      ['v']  = {
        color = colors.visual,
        bold = false,
        italic = false,
      },
      ['V']  = {
        color = colors.visual,
        bold = false,
        italic = false,
      },
      [''] = {
        color = colors.visual,
        bold = false,
        italic = false,
      },
      ['s']  = {
        color = colors.select,
        bold = false,
        italic = false,
      },
      ['S']  = {
        color = colors.select,
        bold = false,
        italic = false,
      },
      ['R']  = {
        color = colors.replace,
        bold = false,
        italic = false,
      },
      ['c']  = {
        color = colors.command,
        bold = false,
        italic = false,
      },
    }
  }
}
```

since in Neovim, the "highlght" includes italic/bold. Or what do you think? Or do you think that would make it more confusing the user to include italic/bold under the `highlights` table?

mawkler commented on lua/modicator/init.lua:
```diff
@@ -24,24 +24,42 @@ local options = {
   local base_highlight = api.nvim_get_hl_by_name('CursorLineNr', true)
-  local opts = vim.tbl_extend('keep', { foreground = color }, base_highlight)
+  -- Should this be defined with these defaults?
+  local base_format = { bold = false, italic = false }
```
These make sense as default, but perhaps we should expose these to the user in the configuration API? I have a feeling that most people want the same formatting for all modes :)

mawkler commented on lua/modicator/init.lua:
```diff
@@ -24,24 +24,42 @@ local options = {
+  local base_format = { bold = false, italic = false }
+  local opts = vim.tbl_extend('keep',
+    { foreground = color, bold = format['bold'], italic = format['italic'] },
+  base_highlight, base_format)
```
I would prefer if this function call got line broken like so:

```lua
local opts = vim.tbl_extend(
  'keep',
  { foreground = color, bold = format['bold'], italic = format['italic'] },
  base_highlight,
  base_format
)
```

rathharper33 committed:
refactor: bold/italic options

rathharper33 committed:
docs: add new config option, installation with lazy.nvim

rathharper33 commented:
A quick summary of my changes:

- To `options`:
  - added the key `overrides`, which one can use to set every mode to bold/italic
  - refactored `options.highlights.modes` as shown in the above comment
- To `M.setup`:
  - new variable `has_overrides` - if the user has given overrides
- To `M.set_format`:
  - rewrote for the new `options` keys, removed the call to `vim.tbl_extend` as it wasn't necessary anymore

I also added the updated `options` keys to config, as well as how to install with [lazy.nvim](https://github.com/folke/lazy.nvim).

mawkler commented:
@itzaeon Nice! However, now there's no highlight except for in normal mode for me. If I go to insert/visual/replace/etc. mode the line number indicator gets the foreground color of `Nomal` (white for me). Do you have the same issue?

mawkler commented on lua/modicator/init.lua:
```diff
@@ -12,36 +12,80 @@ end
+--- Set the foreground color, bold text, and italic text
+--- of 'CursorLineNr'.
+--- @param format table
+M.set_format = function(format)
```
I would prefer if we call the method `M.set_highlight` and its parameter `highlight` (instead of `format`) since Neovim refers to the color and formatting as "highlight" 🙂

rathharper33 commented:
I believe I know why you experience this behavior. The two colorschemes I use are [zephyr](https://github.com/glepnir/zephyr-nvim) and [oxocarbon](https://github.com/nyoom-engineering/oxocarbon.nvim). This plugin works "as designed" when I use `zephyr` as my scheme, but not when I use `oxocarbon`.

Highlights with `zephyr`:
- `CursorLineNr` (normal mode): guifg=#61afef
- `Question` (insert mode): guifg=#f7bb3b
- `Type` (visual mode): guifg=#1abc9c

And with `oxocarbon`:
- `CursorLineNr` (normal mode): guifg=**#d0d0d0**
- `Question` (insert mode): guifg=**#d0d0d0**
- `Type` (visual mode): guifg=#78a9ff

Note that with `oxocarbon`, both `CursorLineNr` and `Question` have the same value. I see the same behavior that you do. With `zephyr`, since CursorLineNr/Question/Type are all different, the plugin works as intended.

I see two options to fix this. 1) We can define highlight groups specifically for modicator (ie ModicatorNormalHighlight, ModicatorInsertHighlight) or 2) a new function can be added that finds *different* highlights for each mode.

mawkler commented:
@itzaeon I'm not sure that's the issue I'm seeing since I explicitly set the colors in my config. I get the same problem with the following test config on your branch:

```lua
use { 'melkster/modicator.nvim',
  config = function()
    require('modicator').setup({
      highlights = {
        modes = {
          ['i']  = 'red',
          ['v']  = 'green',
          ['V']  = 'blue',
          [''] = 'blue',
          ['s']  = 'yellow',
          ['S']  = 'brown',
          ['R']  = 'pink',
          ['c']  = 'orange',
        }
      },
    })
  end
}
```

rathharper33 commented:
@mawkler That config won't work because it was changed [here](https://github.com/mawkler/modicator.nvim/pull/3#issuecomment-1403198215). I get the same behavior using that config, but because the config options have changed in accordance with your comment, something like this should work:

```lua
use { 'melkster/modicator.nvim',
  config = function()
    require('modicator').setup({
      highlights = {
        modes = {
          ['n'] = { color = 'orange' },
          ['i']  = { color = 'red' },
          ['v']  = { color = 'green' },
          ['V']  = { color = 'blue' },
          ['�'] = { color = 'blue' },
          ['s']  = { color = 'yellow' },
          ['S']  = { color = 'brown' },
          ['R']  = { color = 'pink' },
          ['c']  = { color = 'orange' },
        }
      },
    })
  end
}
```

mawkler commented:
@itzaeon Of course! You're right, that's my bad.

mawkler commented on lua/modicator/init.lua:
```diff
@@ -12,36 +12,80 @@ end
   highlights = {
+    overrides = {
+      bold = false,
+      bold = false
```
You accidentally put `bold` twice here

mawkler commented on lua/modicator/init.lua:
```diff
@@ -12,36 +12,80 @@ end
+  local args = {
+    foreground = format.color,
+    bold = has_overrides and ors.bold or format.bold,
+    italic = has_overrides and ors.italic or format.italic
```
You seem to be using a global variable `has_overrides` to determine if the highlight should be bold and/or italic, and global variables should generally be avoided.

Perhaps we should call the option table's default bold/italic options `defaults` instead of `overrides`. Then we can use `vim.tbl_extend()` to merge the defal do something like this instead:

```lua
local args = {
  foreground = format.color,
  bold = format.bold,
  italic = format.italic
}
vim.tbl_extend('force', options.highlights.overrides, args)

rathharper33 committed:
refactor: change name of set_format to set_highlight

rathharper33 committed:
fix: duplicate bold table entry

rathharper33 committed:
remove global has_overrides, use vim.tbl_extend() instead

rathharper33 committed:
fix: use keep instead of force in vim.tbl_extend

rathharper33 commented:
Those should work.

rathharper33 committed:
fix: remove stray print statement

mawkler commented on README.md:
```diff
@@ -50,15 +68,59 @@ local modicator = require('modicator')
+    -- to every mode.
+    overrides = {
+      bold = false,
+      bold = false
```
Don't forget to change this one to `italic` as well :)

mawkler reviewed (commented):
This is looking great! The only thing I want now is for `highlights.overrides` in the options table passed to `setup()` to be called `highlights.defaults` 🙂

rathharper33 committed:
refactor: change highlights.overrides to highlights.defaults

rathharper33 committed:
fix: overrides --> defaults in README.md

rathharper33 commented:
Done!

mawkler commented:
Awesome! Thank you very much for your contributing! 😄

mawkler commented:
@itzaeon Note that I just changed the name of the option `color` to `foreground` in 26338f79329e0659b5c4bd7909816f95a715b6ca to make it more coherent with Neovim's `vim.api.nvim_set_hl()`. I also added a `background` option.
